diesel/src/pg/expression/mod.rs

```diff
@@ -9,6 +9,8 @@ pub(crate) mod array;
 pub mod array_comparison;
 pub(crate) mod expression_methods;
 pub mod extensions;
+#[cfg(not(feature = "sqlite"))]
```

Comment:
Why `not(feature = "sqlite")` and not `feature = "postgres")` as the corresponding feature is postgres specific.

Reply:
This stems from the way the `sql_function!` works, and to make the library compile with both the `postgres` and `sqlite` features enabled. Without, the following error is thrown out:

```
error[E0277]: the trait bound `sqlite::backend::Sqlite: sql_types::HasSqlType<pg::types::sql_types::Cidr>` is not satisfied
  --> diesel/src/pg/expression/functions.rs:43:1
   |
43 | / sql_function! {
44 | |     /// Returns the network part of the address, zeroing out whatever is to the right of the
45 | |     /// netmask. (This is equivalent to casting the value to cidr.)
46 | |     fn network(addr: Inet) -> Cidr;
47 | | }
   | | ^
   | | |
   | |_the trait `sql_types::HasSqlType<pg::types::sql_types::Cidr>` is not implemented for `sqlite::backend::Sqlite`
   |   in this macro invocation
   |
   = help: the following implementations were found:
             <sqlite::backend::Sqlite as sql_types::HasSqlType<sql_types::BigInt>>
             <sqlite::backend::Sqlite as sql_types::HasSqlType<sql_types::Binary>>
             <sqlite::backend::Sqlite as sql_types::HasSqlType<sql_types::Bool>>
             <sqlite::backend::Sqlite as sql_types::HasSqlType<sql_types::Date>>
           and 8 others
```

The gist of it is, `sql_function!` will try to define the function for the SQLite backend regardless of wether it supports it when the `sqlite` feature flag is turned on. When both features are defined (such as in the `bin/test` script), those functions are defined by the Pg-specific code and installed on the SQLite driver.

I am not using `feature = "postgres"` here as the `pg` mod is already behind this feature gate. A more correct way to say it would be "when only postgres is turned on", but I did not bother since this poses no build error with mysql enabled (though it may allow for some erroneous queries at runtime, so it may be a good idea to `not(feature = "sqlite", feature = "mysql")`).

As for the rationale of "not compiling stuff in for multiple backends", I would say it is OK to disable some backend-specific stuff is enabling more than one backend, as you most likely target some cross-backend compatibility by turning multiple ones on.

Reply:
Ok that look interesting. Compiling diesel with more than one backend enabled is definitively something we need to support. In my opinion it is not a solution to just provide don't provide all postgres specific functions in that case, is it is bad design to disable features that should be supported when more backends are enabled. I consider that more a bug of our `sql_function!` implementation that should be fixed there. I've looked a bit into why this is happening, but found no fast fix yet. Therefore I propose the following (as fixing this as part of this PR is probably a lot of work):
* You open a new issue stating that defining that function in a crate enabling the `sqlite` feature and the `postgres` feature will result in the mentioned error.
* We remove this single function (`network`) from the implementation of this PR for now, as this is the only remaining blocker for merging. After all it's quite easy to just define that function in any downstream crate if required. 
* I will try to look into the underlying issue with `sql_function!` and try to come up with a way to fix that behavior.

Reply:
I think this hides a much more deep implementation issue in diesel's `sql_function!` macro. As you noted, the issue only arises for the `network` function, and not the others. The only difference this one has is it takes a concrete type (`Inet`) rather than a generic type (`T: InetOrCird + SingleValue`). Indeed, when using a generic type parameter, it compiles just fine.

Indeed, without the generic parameters, `sql_function!` does generates the `register_impl` and `register_nondeterministic_impl` functions which raise the error. Either this is wanted behaviour (no generics for sqlite), or this is an oversight in the macro.

For the time being, I'll just change `network` to accept `InetOrCidr` like other functions (since the actual postgres function accepts both, this is more correct), and the build should pass fine now.

However, I will still open the issue as this is clearly a bug and the "generics don't work for sqlite" is likely another one.

Reply:
Not accepting generic sql functions for sqlite was an explicit design decision as we need the the concrete types for deserializing the arguments internally. (Custom sql functions in sqlite are implemented by providing a c compatible callback via ffi)

Reply:
Great! Means that switching the function to a generic is a hacky way, but future proof way, to get this to _not_ be installed on the sqlite driver. Seems like we have a solution then (for this PR at least).

I think we are good now.